README.md
# GoAccepts

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/fc86e14746f04abe966d3c771cfb41af)](https://app.codacy.com/app/Svjard/goaccepts?utm_source=github.com&utm_medium=referral&utm_content=Svjard/goaccepts&utm_campaign=Badge_Grade_Dashboard)
[![Go Report Card](https://goreportcard.com/badge/github.com/Svjard/goaccepts)](https://goreportcard.com/report/github.com/Svjard/goaccepts) [![Build Status](https://travis-ci.com/Svjard/goaccepts.svg?branch=master)](https://travis-ci.com/Svjard/goaccepts) [![codecov](https://codecov.io/gh/Svjard/goaccepts/branch/master/graph/badge.svg)](https://codecov.io/gh/Svjard/goaccepts)
[![DeepSource](https://static.deepsource.io/deepsource-badge-light.svg)](https://deepsource.io/gh/Svjard/goaccepts/?ref=repository-badge)

Utility library that can parse the `Accept*` header to pull out encodings/languages/charsets/mimetypes for use by an API.

## Usage

```
import "github.com/Svjard/goaccepts"

// Header sent in request with "en-US, it;q=0.6"
...
goaccepts.Languages(res.headers["Accept-Language"])

// will return []string{"en-US","it"}
```

## Development

Each header type is seperated into its own file and parser.

Accept          - media-type.go  
Accept-Encoding - encoding.go  
Accept-Charset  - charset.go  
Accept-Language - language.go   
  
RFC Reference: https://tools.ietf.org/html/rfc2616 
  
Each module returns has two specific functions, the first returns a raw string array representation of the accepted entities in order of weight. The second returns an array of structures which provides a specific breakdown and detail about each parsed entity.  
  
## Testing

To run the test suite run the following command:

```go test -v ./...```

To run the code coverage report run the following command:

```./go.test.sh```


CHANGELOG.md
# CHANGELOG

### v0.1.0

- Initial release

charset_test.go
package goaccepts_test

import (
	"testing"

	"github.com/Svjard/goaccepts"

	"github.com/stretchr/testify/assert"
)

var testMessageC = "Expect charsets to match parsed header charsets"

func TestCharsets(t *testing.T) {
	assert.Equal(t, goaccepts.Languages(nil), []string{"*"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("*"), []string{"*"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("*, UTF-8"), []string{"*", "UTF-8"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("*, UTF-8;q=0"), []string{"*"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("ISO-8859-1"), []string{"ISO-8859-1"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("UTF-8;q=0"), []string{}, testMessageC)
	assert.Equal(t, goaccepts.Languages("UTF-8, ISO-8859-1"), []string{"UTF-8", "ISO-8859-1"}, testMessageC)
	assert.Equal(t, goaccepts.Languages("UTF-8;q=0.8, ISO-8859-1"), []string{"ISO-8859-1", "UTF-8"}, testMessageC)
}

func testCompareC(a, b []goaccepts.CharsetResult) bool {
	if len(a) != len(b) {
		return false
	}

	for i := 0; i < len(a); i++ {
		if a[i].Charset != b[i].Charset {
			return false
		}

		if a[i].Weight != b[i].Weight {
			return false
		}
	}

	return true
}

func TestCharsetDetails(t *testing.T) {
	asterisk := goaccepts.CharsetResult{
		Charset: "*",
		Weight:  100,
	}
	utf8 := goaccepts.CharsetResult{
		Charset: "UTF-8",
		Weight:  1.0,
	}
	iso := goaccepts.CharsetResult{
		Charset: "ISO-8859-1",
		Weight:  1.0,
	}
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails(nil), []goaccepts.CharsetResult{asterisk}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("*"), []goaccepts.CharsetResult{asterisk}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("*, UTF-8"), []goaccepts.CharsetResult{asterisk, utf8}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("*, UTF-8;q=0"), []goaccepts.CharsetResult{asterisk}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("ISO-8859-1"), []goaccepts.CharsetResult{iso}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("UTF-8;q=0"), []goaccepts.CharsetResult{}), true, testMessageC)
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("UTF-8, ISO-8859-1"), []goaccepts.CharsetResult{utf8, iso}), true, testMessageC)
	utf88 := goaccepts.CharsetResult{
		Charset: "UTF-8",
		Weight:  0.8,
	}
	assert.Equal(t, testCompareC(goaccepts.CharsetsDetails("UTF-8;q=0.8, ISO-8859-1"), []goaccepts.CharsetResult{iso, utf88}), true, testMessageC)
}


encoding_test.go
package goaccepts_test

import (
	"testing"

	"github.com/Svjard/goaccepts"

	"github.com/stretchr/testify/assert"
)

var testMessageE = "Expect encodings to match parsed header encodings"

func TestEncodings(t *testing.T) {
	assert.Equal(t, goaccepts.Encodings(nil), []string{"identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("*"), []string{"*", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("*, gzip"), []string{"*", "gzip", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("*, gzip;q=0"), []string{"*", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("*;q=0"), []string{}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("*;q=0, identity;q=1"), []string{"identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("identity"), []string{"identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("identity;q=0"), []string{}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("gzip"), []string{"gzip", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("gzip, compress;q=0"), []string{"gzip", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("gzip, deflate"), []string{"gzip", "deflate", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("gzip;q=0.8, deflate"), []string{"deflate", "gzip", "identity"}, testMessageE)
	assert.Equal(t, goaccepts.Encodings("gzip;q=0.8, identity;q=0.5, *;q=0.3"), []string{"gzip", "identity", "*"}, testMessageE)
}

func testCompareE(a, b []goaccepts.EncodingResult) bool {
	if len(a) != len(b) {
		return false
	}

	for i := 0; i < len(a); i++ {
		if a[i].Encoding != b[i].Encoding {
			return false
		}

		if a[i].Weight != b[i].Weight {
			return false
		}
	}

	return true
}

func TestEncodingsDetails(t *testing.T) {
	asterisk := goaccepts.EncodingResult{
		Encoding: "*",
		Weight:   100,
	}
	gzip := goaccepts.EncodingResult{
		Encoding: "gzip",
		Weight:   1.0,
	}
	deflate := goaccepts.EncodingResult{
		Encoding: "deflate",
		Weight:   1.0,
	}
	identity := goaccepts.EncodingResult{
		Encoding: "identity",
		Weight:   1.0,
	}
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails(nil), []goaccepts.EncodingResult{identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("*"), []goaccepts.EncodingResult{asterisk, identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("*, gzip"), []goaccepts.EncodingResult{asterisk, gzip, identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("*, gzip;q=0"), []goaccepts.EncodingResult{asterisk, identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("*;q=0"), []goaccepts.EncodingResult{}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("*;q=0, identity;q=1"), []goaccepts.EncodingResult{identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("identity"), []goaccepts.EncodingResult{identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("identity;q=0"), []goaccepts.EncodingResult{}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("gzip"), []goaccepts.EncodingResult{gzip, identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("gzip, compress;q=0"), []goaccepts.EncodingResult{gzip, identity}), true, testMessageE)
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("gzip, deflate"), []goaccepts.EncodingResult{gzip, deflate, identity}), true, testMessageE)

	gzip8 := goaccepts.EncodingResult{
		Encoding: "gzip",
		Weight:   0.8,
	}
	identity8 := goaccepts.EncodingResult{
		Encoding: "identity",
		Weight:   0.8,
	}
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("gzip;q=0.8, deflate"), []goaccepts.EncodingResult{deflate, gzip8, identity8}), true, testMessageE)
	identity5 := goaccepts.EncodingResult{
		Encoding: "identity",
		Weight:   0.5,
	}
	asterisk3 := goaccepts.EncodingResult{
		Encoding: "*",
		Weight:   0.3,
	}
	assert.Equal(t, testCompareE(goaccepts.EncodingsDetails("gzip;q=0.8, identity;q=0.5, *;q=0.3"), []goaccepts.EncodingResult{gzip8, identity5, asterisk3}), true, testMessageE)
}


language_test.go
package goaccepts_test

import (
	"testing"

	"github.com/Svjard/goaccepts"

	"github.com/stretchr/testify/assert"
)

var testMessageL = "Expect languages to match parsed header languages"

func TestLanguages(t *testing.T) {
	assert.Equal(t, goaccepts.Languages(nil), []string{"*"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("*"), []string{"*"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("*, en"), []string{"*", "en"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("*, en;q=0"), []string{"*"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("*;q=0.8, en, es"), []string{"en", "es", "*"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en"), []string{"en"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en;q=0"), []string{}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en;q=0.8, es"), []string{"es", "en"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en;q=0.9, es;q=0.8, en;q=0.7"), []string{"en", "es"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en-US, en;q=0.8"), []string{"en-US", "en"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en-US, en-GB"), []string{"en-US", "en-GB"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("en-US;q=0.8, es"), []string{"es", "en-US"}, testMessageL)
	assert.Equal(t, goaccepts.Languages("nl;q=0.5, fr, de, en, it, es, pt, no, se, fi, ro"), []string{"fr", "de", "en", "it", "es", "pt", "no", "se", "fi", "ro", "nl"}, testMessageL)
}

func testCompareL(a, b []goaccepts.LanguageResult) bool {
	if len(a) != len(b) {
		return false
	}

	for i := 0; i < len(a); i++ {
		if a[i].Prefix != b[i].Prefix {
			return false
		}

		if a[i].Suffix != b[i].Suffix {
			return false
		}

		if a[i].Weight != b[i].Weight {
			return false
		}

		if a[i].Full != b[i].Full {
			return false
		}
	}

	return true
}

func TestLanguageDetails(t *testing.T) {
	asterisk := goaccepts.LanguageResult{
		Prefix: "*",
		Suffix: "",
		Weight: 100,
		Full:   "*",
	}
	en := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "",
		Weight: 1.0,
		Full:   "en",
	}
	es := goaccepts.LanguageResult{
		Prefix: "es",
		Suffix: "",
		Weight: 1.0,
		Full:   "es",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails(nil), []goaccepts.LanguageResult{asterisk}), true, testMessageL)
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("*"), []goaccepts.LanguageResult{asterisk}), true, testMessageL)
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("*, en"), []goaccepts.LanguageResult{asterisk, en}), true, testMessageL)
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("*, en;q=0"), []goaccepts.LanguageResult{asterisk}), true, testMessageL)

	asterisk8 := goaccepts.LanguageResult{
		Prefix: "*",
		Suffix: "",
		Weight: 0.8,
		Full:   "*",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("*;q=0.8, en, es"), []goaccepts.LanguageResult{en, es, asterisk8}), true, testMessageL)
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en"), []goaccepts.LanguageResult{en}), true, testMessageL)
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en;q=0"), []goaccepts.LanguageResult{}), true, testMessageL)
	en8 := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "",
		Weight: 0.8,
		Full:   "en",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en;q=0.8, es"), []goaccepts.LanguageResult{es, en8}), true, testMessageL)

	en9 := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "",
		Weight: 0.9,
		Full:   "en",
	}
	es8 := goaccepts.LanguageResult{
		Prefix: "es",
		Suffix: "",
		Weight: 0.8,
		Full:   "es",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en;q=0.9, es;q=0.8, en;q=0.7"), []goaccepts.LanguageResult{en9, es8}), true, testMessageL)
	enUs := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "US",
		Weight: 1.0,
		Full:   "en-US",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en-US, en;q=0.8"), []goaccepts.LanguageResult{enUs, en8}), true, testMessageL)
	enGb := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "GB",
		Weight: 1.0,
		Full:   "en-GB",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en-US, en-GB"), []goaccepts.LanguageResult{enUs, enGb}), true, testMessageL)
	enUs8 := goaccepts.LanguageResult{
		Prefix: "en",
		Suffix: "US",
		Weight: 0.8,
		Full:   "en-US",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("en-US;q=0.8, es"), []goaccepts.LanguageResult{es, enUs8}), true, testMessageL)
	nl5 := goaccepts.LanguageResult{
		Prefix: "nl",
		Suffix: "",
		Weight: 0.5,
		Full:   "nl",
	}
	fr := goaccepts.LanguageResult{
		Prefix: "fr",
		Suffix: "",
		Weight: 1.0,
		Full:   "fr",
	}
	de := goaccepts.LanguageResult{
		Prefix: "de",
		Suffix: "",
		Weight: 1.0,
		Full:   "de",
	}
	it := goaccepts.LanguageResult{
		Prefix: "it",
		Suffix: "",
		Weight: 1.0,
		Full:   "it",
	}
	pt := goaccepts.LanguageResult{
		Prefix: "pt",
		Suffix: "",
		Weight: 1.0,
		Full:   "pt",
	}
	no := goaccepts.LanguageResult{
		Prefix: "no",
		Suffix: "",
		Weight: 1.0,
		Full:   "no",
	}
	se := goaccepts.LanguageResult{
		Prefix: "se",
		Suffix: "",
		Weight: 1.0,
		Full:   "se",
	}
	fi := goaccepts.LanguageResult{
		Prefix: "fi",
		Suffix: "",
		Weight: 1.0,
		Full:   "fi",
	}
	ro := goaccepts.LanguageResult{
		Prefix: "ro",
		Suffix: "",
		Weight: 1.0,
		Full:   "ro",
	}
	assert.Equal(t, testCompareL(goaccepts.LanguagesDetails("nl;q=0.5, fr, de, en, it, es, pt, no, se, fi, ro"), []goaccepts.LanguageResult{fr, de, en, it, es, pt, no, se, fi, ro, nl5}), true, testMessageL)
}


charset.go
package goaccepts

import (
	"sort"
	"strconv"
	"strings"
)

// CharsetResult of the accepts header
type CharsetResult struct {
	// Charset ...
	Charset string
	// Weight ...
	Weight float64
	// i ...
	i int
}

func parseAcceptCharset(accept string) []CharsetResult {
	res := []CharsetResult{}
	accepts := strings.Split(accept, ",")

	j := 0
	for i := 0; i < len(accepts); i++ {
		charset := parseCharset(strings.Trim(accepts[i], " "), i)
		if charset != nil {
			n := findC(res, charset.Charset)
			if n == -1 {
				res = append(res[:j], append([]CharsetResult{*charset}, res[j:]...)...)
				j = j + 1
			}
		}
	}

	return res
}

func parseCharset(str string, i int) *CharsetResult {
	if str == "*" {
		return &CharsetResult{
			Charset: str,
			Weight:  100.0,
			i:       i,
		}
	}

	weight := 1.0
	params := strings.Split(str, ";")
	if len(params) > 1 {
		p := strings.Split(params[1], "=")
		if strings.ToLower(p[0]) == "q" {
			weight, _ = strconv.ParseFloat(p[1], 64)
		}
	}

	charset := params[0]
	return &CharsetResult{
		Charset: charset,
		Weight:  weight,
		i:       i,
	}
}

func parseCharsets(accept interface{}) []CharsetResult {
	// no header = *
	val := "*"
	if accept != nil {
		val = accept.(string)
	}

	accepts := parseAcceptCharset(val)

	// sorted list of all languages
	filteredList := filterC(accepts, func(spec CharsetResult) bool {
		return spec.Weight > 0
	})
	sort.SliceStable(filteredList, func(i, j int) bool {
		return filteredList[i].Weight > filteredList[j].Weight
	})

	return filteredList
}

// Charsets - Get the list of accepted charsets from an Accept-Charset header.
// https://tools.ietf.org/html/rfc2616#section-14.4
func Charsets(accept interface{}) []string {
	charsets := parseCharsets(accept)

	fullList := mapC(charsets, func(spec CharsetResult) string {
		return spec.Charset
	})

	return fullList
}

// CharsetsDetails - Get the list of accepted charsets from an Accept-Charset header.
// https://tools.ietf.org/html/rfc2616#section-14.2
func CharsetsDetails(accept interface{}) []CharsetResult {
	return parseCharsets(accept)
}

func filterC(vs []CharsetResult, f func(CharsetResult) bool) []CharsetResult {
	vsf := make([]CharsetResult, 0)
	for _, v := range vs {
		if f(v) {
			vsf = append(vsf, v)
		}
	}
	return vsf
}

func mapC(vs []CharsetResult, f func(CharsetResult) string) []string {
	vsf := make([]string, 0)
	for _, v := range vs {
		vsf = append(vsf, f(v))
	}
	return vsf
}

func findC(a []CharsetResult, x string) int {
	for i, n := range a {
		if x == n.Charset {
			return i
		}
	}
	return -1
}


language.go
package goaccepts

import (
	"sort"
	"strconv"
	"strings"
)

// LanguageResult of the accepts header
type LanguageResult struct {
	// prefix ...
	Prefix string
	// suffix ...
	Suffix string
	// weight ...
	Weight float64
	// i ...
	i int
	// full ...
	Full string
}

func parseAcceptLanguage(accept string) []LanguageResult {
	res := []LanguageResult{}
	accepts := strings.Split(accept, ",")

	j := 0
	for i := 0; i < len(accepts); i++ {
		language := parseLanguage(strings.Trim(accepts[i], " "), i)
		if language != nil {
			n := findL(res, language.Full)
			if n == -1 {
				res = append(res[:j], append([]LanguageResult{*language}, res[j:]...)...)
				j = j + 1
			}
		}
	}

	return res
}

func parseLanguage(str string, i int) *LanguageResult {
	if str == "*" {
		return &LanguageResult{
			Prefix: str,
			Suffix: "",
			Weight: 100.0,
			i:      i,
			Full:   str,
		}
	}

	weight := 1.0
	params := strings.Split(str, ";")
	if len(params) > 1 {
		p := strings.Split(params[1], "=")
		if strings.ToLower(p[0]) == "q" {
			weight, _ = strconv.ParseFloat(p[1], 64)
		}
	}

	langs := strings.Split(params[0], "-")
	prefix := langs[0]
	suffix := ""
	if len(langs) > 1 {
		suffix = strings.Join(langs[1:], "-")
	}
	full := params[0]

	return &LanguageResult{
		Prefix: prefix,
		Suffix: suffix,
		Weight: weight,
		i:      i,
		Full:   full,
	}
}

func parseLanguages(accept interface{}) []LanguageResult {
	// no header = *
	val := "*"
	if accept != nil {
		val = accept.(string)
	}

	accepts := parseAcceptLanguage(val)

	// sorted list of all languages
	filteredList := filterL(accepts, func(spec LanguageResult) bool {
		return spec.Weight > 0
	})
	sort.SliceStable(filteredList, func(i, j int) bool {
		return filteredList[i].Weight > filteredList[j].Weight
	})

	return filteredList
}

// Languages - Get the preferred languages from an Accept-Language header.
// https://tools.ietf.org/html/rfc2616#section-14.4
func Languages(accept interface{}) []string {
	languages := parseLanguages(accept)

	fullList := mapL(languages, func(spec LanguageResult) string {
		return spec.Full
	})

	return fullList
}

// LanguagesDetails - Get the preferred languages from an Accept-Language header.
// https://tools.ietf.org/html/rfc2616#section-14.4
func LanguagesDetails(accept interface{}) []LanguageResult {
	return parseLanguages(accept)
}

func filterL(vs []LanguageResult, f func(LanguageResult) bool) []LanguageResult {
	vsf := make([]LanguageResult, 0)
	for _, v := range vs {
		if f(v) {
			vsf = append(vsf, v)
		}
	}
	return vsf
}

func mapL(vs []LanguageResult, f func(LanguageResult) string) []string {
	vsf := make([]string, 0)
	for _, v := range vs {
		vsf = append(vsf, f(v))
	}
	return vsf
}

func findL(a []LanguageResult, x string) int {
	for i, n := range a {
		if x == n.Full {
			return i
		}
	}
	return -1
}


mime-type.go
package goaccepts

import (
	"sort"
	"strconv"
	"strings"
)

// MimeTypeResult of the accepts header
type MimeTypeResult struct {
	// Type ...
	Type string
	// ParentType ...
	ParentType string
	// ChildType
	ChildType string
	// Level ...
	Level int64
	// Quality ...
	Quality float64
	// i ...
	i int
}

func parseAcceptMimeType(accept string) []MimeTypeResult {
	res := []MimeTypeResult{}
	accepts := strings.Split(accept, ",")

	j := 0
	for i := 0; i < len(accepts); i++ {
		mimeType := parseMimeType(strings.Trim(accepts[i], " "), i)
		if mimeType != nil {
			n := findM(res, mimeType.Type)
			if n == -1 {
				res = append(res[:j], append([]MimeTypeResult{*mimeType}, res[j:]...)...)
				j = j + 1
			}
		}
	}

	return res
}

func parseMimeType(str string, i int) *MimeTypeResult {
	if str == "*/*" {
		return &MimeTypeResult{
			Type:       str,
			ParentType: "*",
			ChildType:  "*",
			Quality:    100.0,
			i:          i,
		}
	}

	quality := 1.0
	var level int64 = int64(0)
	params := strings.Split(str, ";")
	mimeType := strings.ToLower(params[0])
	if strings.IndexByte(mimeType, '"') == 0 && strings.LastIndexByte(mimeType, '"') == len(mimeType)-1 {
		mimeType = strings.Trim(mimeType, "\"")
	}
	typeParts := strings.Split(mimeType, "/")
	childType := ""
	if len(typeParts) > 1 {
		childType = typeParts[1]
	}

	if len(params) > 1 {
		for n := 1; n < len(params); n++ {
			p := strings.Split(params[n], "=")
			if p[0] == "q" {
				quality, _ = strconv.ParseFloat(p[1], 64)
			}

			if strings.ToLower(p[0]) == "level" {
				level, _ = strconv.ParseInt(p[1], 10, 64)
			}
		}
	}

	if level > 0 {
		return &MimeTypeResult{
			Type:       mimeType,
			ParentType: typeParts[0],
			ChildType:  childType,
			Level:      level,
			Quality:    quality,
			i:          i,
		}
	}

	return &MimeTypeResult{
		Type:       mimeType,
		ParentType: typeParts[0],
		ChildType:  childType,
		Quality:    quality,
		i:          i,
	}
}

func parseMimeTypes(accept interface{}) []MimeTypeResult {
	// no header = */*
	val := "*/*"
	if accept != nil {
		val = accept.(string)
	}

	accepts := parseAcceptMimeType(val)

	// sorted list of all languages
	filteredList := filterM(accepts, func(spec MimeTypeResult) bool {
		return spec.Quality > 0
	})
	sort.SliceStable(filteredList, func(i, j int) bool {
		return filteredList[i].Quality > filteredList[j].Quality
	})

	return filteredList
}

// MimeTypes - Get the list of accepted mime types from an Accept header.
// https://tools.ietf.org/html/rfc2616#section-14.1
func MimeTypes(accept interface{}) []string {
	charsets := parseMimeTypes(accept)

	fullList := mapM(charsets, func(spec MimeTypeResult) string {
		return spec.Type
	})

	return fullList
}

// MimeTypesDetails - Get the list of accepted mime types from an Accept header.
// https://tools.ietf.org/html/rfc2616#section-14.1
func MimeTypesDetails(accept interface{}) []MimeTypeResult {
	return parseMimeTypes(accept)
}

func filterM(vs []MimeTypeResult, f func(MimeTypeResult) bool) []MimeTypeResult {
	vsf := make([]MimeTypeResult, 0)
	for _, v := range vs {
		if f(v) {
			vsf = append(vsf, v)
		}
	}
	return vsf
}

func mapM(vs []MimeTypeResult, f func(MimeTypeResult) string) []string {
	vsf := make([]string, 0)
	for _, v := range vs {
		vsf = append(vsf, f(v))
	}
	return vsf
}

func findM(a []MimeTypeResult, x string) int {
	for i, n := range a {
		if x == n.Type {
			return i
		}
	}
	return -1
}


encoding.go
package goaccepts

import (
	"math"
	"sort"
	"strconv"
	"strings"
)

// EncodingResult of the accepts header
type EncodingResult struct {
	// Encoding ...
	Encoding string
	// weight ...
	Weight float64
	// i ...
	i int
}

func identityCheck(results []EncodingResult, encoding *EncodingResult) bool {
	index := findE(results, "*")

	if strings.ToLower(encoding.Encoding) == "identity" {
		return true
	}

	if index != -1 && results[index].Encoding == "*" && results[index].Weight == 0 {
		return true
	}

	return false
}

func parseAcceptEncoding(accept string) []EncodingResult {
	res := []EncodingResult{}
	accepts := strings.Split(accept, ",")
	hasIdentity := false
	minQuality := 1.0

	j := 0
	for i := 0; i < len(accepts); i++ {
		encoding := parseEncoding(strings.Trim(accepts[i], " "), i)
		if encoding != nil {
			n := findE(res, encoding.Encoding)
			if n == -1 {
				res = append(res[:j], append([]EncodingResult{*encoding}, res[j:]...)...)
				hasIdentity = hasIdentity || identityCheck(res, encoding)
				minQuality = math.Min(minQuality, encoding.Weight)
				j = j + 1
			}
		}
	}

	if hasIdentity == false {
		/*
		 * If identity doesn't explicitly appear in the accept-encoding header,
		 * it's added to the list of acceptable encoding with the lowest q
		 */
		identWeight := minQuality
		if minQuality == 0 {
			identWeight = 1.0
		}
		ident := &EncodingResult{
			Encoding: "identity",
			Weight:   identWeight,
			i:        j + 1,
		}
		res = append(res[:j], append([]EncodingResult{*ident}, res[j:]...)...)
	}

	return res
}

func parseEncoding(str string, i int) *EncodingResult {
	if str == "*" {
		return &EncodingResult{
			Encoding: str,
			Weight:   100.0,
			i:        i,
		}
	}

	weight := 1.0
	params := strings.Split(str, ";")
	if len(params) > 1 {
		p := strings.Split(params[1], "=")
		if strings.ToLower(p[0]) == "q" {
			weight, _ = strconv.ParseFloat(p[1], 64)
		}
	}

	encods := strings.Split(params[0], "-")
	return &EncodingResult{
		Encoding: encods[0],
		Weight:   weight,
		i:        i,
	}
}

func parseEncodings(accept interface{}) []EncodingResult {
	// no header = *
	val := "identity"
	if accept != nil {
		val = accept.(string)
	}

	accepts := parseAcceptEncoding(val)

	// sorted list of all languages
	filteredList := filterE(accepts, func(spec EncodingResult) bool {
		return spec.Weight > 0
	})
	sort.SliceStable(filteredList, func(i, j int) bool {
		return filteredList[i].Weight > filteredList[j].Weight
	})

	return filteredList
}

// Encodings - Get the encording from an Accept-Encoding header.
// https://tools.ietf.org/html/rfc2616#section-14.2
func Encodings(accept interface{}) []string {
	encodings := parseEncodings(accept)

	fullList := mapE(encodings, func(spec EncodingResult) string {
		return spec.Encoding
	})

	return fullList
}

// EncodingsDetails - Get the encording from an Accept-Encoding header.
// https://tools.ietf.org/html/rfc2616#section-14.2
func EncodingsDetails(accept interface{}) []EncodingResult {
	return parseEncodings(accept)
}

func filterE(vs []EncodingResult, f func(EncodingResult) bool) []EncodingResult {
	vsf := make([]EncodingResult, 0)
	for _, v := range vs {
		if f(v) {
			vsf = append(vsf, v)
		}
	}
	return vsf
}

func mapE(vs []EncodingResult, f func(EncodingResult) string) []string {
	vsf := make([]string, 0)
	for _, v := range vs {
		vsf = append(vsf, f(v))
	}
	return vsf
}

func findE(a []EncodingResult, x string) int {
	for i, n := range a {
		if x == n.Encoding {
			return i
		}
	}
	return -1
}
